service/raft/raft_group0.cc

```diff
@@ -587,6 +739,503 @@ persistent_discovery::persistent_discovery(raft::server_address self, const peer
+            co_await sleep_abortable(retry_period, as);
+            continue;
+        }
+
+        rslog.info(
+            "upgrade_to_group0(): group 0 configuration does not contain all peers yet."
+            " Missing peers: {}. Current group 0 config: {}. Current group 0 config addresses: {}. Sleeping for a while before retrying...",
+            missing_peers, group0_config, current_config);
+
+        co_await sleep_abortable(retry_period, as);
+    }
+}
+
+// Check if anyone entered `use_new_procedures`.
+// This is a best-effort single round-trip check; we don't retry if some nodes fail to answer.
+static future<bool> anyone_finished_upgrade(
+        const raft::server& group0_server, netw::messaging_service& ms, abort_source& as) {
+    static constexpr auto max_concurrency = 10;
+
+    auto current_config = get_raft_members_inet_addrs(group0_server.get_configuration().current);
+    bool finished = false;
+    co_await max_concurrent_for_each(current_config, max_concurrency, [&] (const gms::inet_address& node) -> future<> {
+        try {
+            // TODO: timeout?
+            auto state = co_await send_get_group0_upgrade_state(ms, node, as);
+            if (state == group0_upgrade_state::use_new_procedures) {
+                finished = true;
+            }
+        } catch (abort_requested_exception&) {
+            throw;
+        } catch (...) {
+            // TODO: are there possible fatal errors which should cause us to fail?
+            rslog.warn("anyone_finished_upgrade: `send_get_group0_upgrade_state({})` failed: {}", node, std::current_exception());
+        }
+    });
+    co_return finished;
+}
+
+// Check if it's possible to reach everyone through `get_group0_upgrade_state` RPC.
+static future<> check_remote_group0_upgrade_state_dry_run(
+        const raft::server& group0_server, netw::messaging_service& ms, abort_source& as) {
+    static constexpr auto max_retry_period = std::chrono::seconds{16};
+    static constexpr auto max_concurrency = 10;
+
+    auto retry_period = std::chrono::seconds{1};
+    while (true) {
+        // Note: we strive to get a response from everyone in a 'single round-trip',
+        // so we don't skip nodes which responded in earlier iterations.
+        // We contact everyone in each iteration even if some of these guys already answered.
+        // We fetch the config again on every attempt to handle the possibility of leaving nodes.
+        auto current_config = get_raft_members_inet_addrs(group0_server.get_configuration().current);
+
+        bool retry = false;
+        co_await max_concurrent_for_each(current_config, max_concurrency, [&] (const gms::inet_address& node) -> future<> {
+            try {
+                rslog.info("check_remote_group0_upgrade_state_dry_run: `send_get_group0_upgrade_state({})`", node);
+                // TODO: timeout?
```

Comment:
You mean the `send_get_group0_upgrade_state` will hang? Won't this cause something like "tcp closed connection" error?

Reply:
If there is no tcp stack to answer your SYN it will take minutes until you will get the error.

Reply:
Ok. But then if we timeout, we'll just retry again, so is there a point in timing out?
Maybe there is because we may remove the member from group 0 config to unstuck upgrade